src/main/java/com/google/devtools/build/lib/skyframe/ActionMetadataHandler.java

```diff
@@ -600,6 +600,12 @@ private static FileArtifactValue fileArtifactValueFromArtifact(
     checkState(!artifact.isTreeArtifact() && !artifact.isMiddlemanArtifact(), artifact);
 
     Path pathNoFollow = artifactPathResolver.toPath(artifact);
+    // Handle the case where a symlink is expected before stat-ing the output so that the case where
+    // it incorrectly is a file triggers the expected NotASymlinkException.
+    if (artifact.isSymlink()) {
+      return FileArtifactValue.createForUnresolvedSymlink(pathNoFollow);
```

Comment:
I think that comment is not useful as is (it talks about the history of the code more than the current state). What I would say is that we start by eagerly running readSymbolicLink to avoid an unnecessary stat in the correct case. My point is that stat could be used to detect whether that is a symbolic link, it just happened not to be used this way. Your approach is nice in that it saves IO (stat then readlink vs just readlink and analyze errors)--I think that is the clue here.

Reply:
Good point, modified the comment.